pdf/model/fonts/ttfparser.go

```diff
@@ -258,86 +369,345 @@ func (t *ttfParser) ParseCmap() (err error) {
 	return
 }
 
+// parseCmapSubtable10 parses information from an (1,0) subtable (symbol).
+func (t *ttfParser) parseCmapSubtable10(offset10 int64) error {
+
+	if t.rec.Chars == nil {
+		t.rec.Chars = make(map[uint16]uint16)
+	}
+
+	t.f.Seek(int64(t.tables["cmap"])+offset10, os.SEEK_SET)
+	var length, language uint32
+	format := t.ReadUShort()
+	if format < 8 {
+		length = uint32(t.ReadUShort())
+		language = uint32(t.ReadUShort())
+	} else {
+		t.ReadUShort()
+		length = t.ReadULong()
+		language = t.ReadULong()
+	}
+	common.Log.Debug("parseCmapSubtable10: format=%d length=%d language=%d",
+		format, length, language)
+
+	if format != 0 {
+		return errors.New("unsupported cmap subtable format")
+	}
+
+	dataStr, err := t.ReadStr(256)
+	if err != nil {
+		return err
+	}
+	data := []byte(dataStr)
+
+	for code, glyphId := range data {
+		t.rec.Chars[uint16(code)] = uint16(glyphId)
+		if glyphId != 0 {
+			fmt.Printf("\t0x%02x ➞ 0x%02x=%c\n", code, glyphId, rune(glyphId))
+		}
+	}
+	return nil
+}
+
+// ParseCmap parses the cmap table in a TrueType font.
+func (t *ttfParser) ParseCmap() (err error) {
+	var offset int64
+	if err = t.Seek("cmap"); err != nil {
+		return
+	}
+	common.Log.Debug("ParseCmap")
+	/* version := */ t.ReadUShort()
+	numTables := int(t.ReadUShort())
+	offset10 := int64(0)
+	offset31 := int64(0)
+	for j := 0; j < numTables; j++ {
+		platformID := t.ReadUShort()
+		encodingID := t.ReadUShort()
+		offset = int64(t.ReadULong())
+		if platformID == 3 && encodingID == 1 {
+			// (3,1) subtable. Windows Unicode.
+			offset31 = offset
+		}
+		//fmt.Printf("(%d,%d) subtable @ %d\n", platformID, encodingID, offset)
+	}
+
+	// Latin font support based on (3,1) table encoding.
+	if offset31 != 0 {
+		err = t.parseCmapSubtable31(offset31)
+		if err != nil {
+			return
+		}
+	}
+
+	// Many non-Latin fonts (including asian fonts) use subtable (1,0).
+
+	if offset10 != 0 {
+		// fmt.Printf("Offset10: %d\n", offset10)
+		err = t.parseCmapVersion(offset10)
+		if err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+func (t *ttfParser) parseCmapVersion(offset int64) error {
+	common.Log.Trace("parseCmapVersion: offset=%d", offset)
+
+	if t.rec.Chars == nil {
+		t.rec.Chars = make(map[uint16]uint16)
+	}
+
+	t.f.Seek(int64(t.tables["cmap"])+offset, os.SEEK_SET)
+	var length, language uint32
+	format := t.ReadUShort()
+	if format < 8 {
+		length = uint32(t.ReadUShort())
+		language = uint32(t.ReadUShort())
+	} else {
+		t.ReadUShort()
+		length = t.ReadULong()
+		language = t.ReadULong()
+	}
+	common.Log.Debug("parseCmapVersion: format=%d length=%d language=%d",
+		format, length, language)
+
+	switch format {
+	case 0:
+		return t.parseCmapFormat0()
+	case 6:
+		return t.parseCmapFormat6()
+	default:
+		common.Log.Debug("ERROR: Unsupported cmap format=%d", format)
+		return nil // XXX: Can't return an error here if creator_test.go is to pass.
```

Comment:
Interesting.  Is that some problem with the font file we are using?

Reply:
Either that or the parser is not general enough. I only see this on this test. Were you using unusual fonts?

Reply:
Sounds like the comment came from me (via compositefonts).  I am trying to review as if reading for the first time.
We will keep the comment and look into this later (especially if we find more cases).
Yes probably an issue with the specific font file used.